Incompatible with ghc8.8 due to MonadFail

```
% stack --resolver lts-15.11 build ztar

Error: While constructing the build plan, the following exceptions were encountered:

In the dependencies for ztar-1.0.1:
    path-0.7.0 from stack configuration does not match >=0.5 && <0.7  (latest matching version is 0.6.1)
    zip-1.3.2 from stack configuration does not match >=1.0 && <1.3  (latest matching version is 1.2.0)
needed since ztar is a build target.

Some different approaches to resolving this:

  * Set 'allow-newer: true'
    in /Users/liam/.stack/config.yaml to ignore all version constraints and build anyway.

  * Recommended action: try adding the following to your extra-deps
    in /Users/liam/.stack/global-project/stack.yaml:

- path-0.6.1@sha256:7a9a56f767a8e47b361af855fea0b4a8d18e7b029734957a65dcd5c0c1abef16,2423
- zip-1.2.0@sha256:b7265f229c47a47c3ddf7e3334c3d0811c35e46a2a4bbae7b532541802445325,4458

Plan construction failed.
```

Added these to my global stack.yaml, then
```
% stack --resolver lts-15.11 build ztar
...
ztar                 > configure
ztar                 > Configuring ztar-1.0.1...
ztar                 > build
ztar                 > Preprocessing library for ztar-1.0.1..
ztar                 > Building library for ztar-1.0.1..
ztar                 > 
ztar                 > on the commandline: warning:
ztar                 >     -Wnoncanonical-monadfail-instances is deprecated: fail is no longer a method of Monad
ztar                 > [1 of 6] Compiling Codec.Archive.Tar
ztar                 > [2 of 6] Compiling Codec.Archive.ZTar.GZip
ztar                 > [3 of 6] Compiling Codec.Archive.ZTar.Tar
ztar                 > [4 of 6] Compiling Codec.Archive.ZTar.Zip
ztar                 > 
ztar                 > /private/var/folders/b4/vndwlkd168l7lnn9zcq_3y080000gn/T/stack-f426e50fcdcf04a2/ztar-1.0.1/src/Codec/Archive/ZTar/Zip.hs:68:24: error:
ztar                 >     • No instance for (MonadFail Zip.ZipArchive)
ztar                 >         arising from a use of ‘fail’
ztar                 >     • In the expression: fail $ "Path does not exist: " ++ path
ztar                 >       In a stmt of a 'do' block:
ztar                 >         if | isFile -> insertFile path
ztar                 >            | isDir -> insertDir path
ztar                 >            | otherwise -> fail $ "Path does not exist: " ++ path
ztar                 >       In the expression:
ztar                 >         do isFile <- liftIO $ doesFileExist path
ztar                 >            isDir <- liftIO $ doesDirectoryExist path
ztar                 >            if | isFile -> insertFile path
ztar                 >               | isDir -> insertDir path
ztar                 >               | otherwise -> fail $ "Path does not exist: " ++ path
ztar                 >    |
ztar                 > 68 |         | otherwise -> fail $ "Path does not exist: " ++ path
ztar                 >    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ztar                 > 
Completed 43 action(s).
```

So we cannot fail within the ZipArchive monad. @brandonchinn178 what would you like done to fix this, there are many possibilities?

--- a/package.yaml
+++ b/package.yaml
@@ -36,10 +36,11 @@ library:
     base: '>= 4.7 && < 5'
     bytestring: '>= 0.10.8 && < 0.11'
     directory: '>= 1.3 && < 1.4'
+    extra:
     filepath: '>= 1.4.1 && < 1.5'
-    path: '>= 0.5 && < 0.7'
+    path: '>= 0.5 && < 0.8'
     # tar: '>= 0.5 && < 0.6'
-    zip: '>= 1.0 && < 1.3'
+    zip: '>= 1.0 && < 1.4'
     zlib: '>= 0.6 && < 0.7'
     # For Codec.Archive.Tar
     deepseq:

--- a/package.yaml
+++ b/package.yaml
@@ -28,6 +28,8 @@ when:
       - -Wincomplete-record-updates
       - -Wincomplete-uni-patterns
       - -Wnoncanonical-monad-instances
+  - condition: impl(ghc <= 8.6)
+    ghc-options:
       - -Wnoncanonical-monadfail-instances
 
 library:
@@ -69,11 +71,11 @@ tests:
     dependencies:
       base:
       bytestring:
-      bytestring-arbitrary:
       filepath:
       path:
       path-io:
       QuickCheck:
+      quickcheck-instances:
       tasty:
       tasty-quickcheck:
       ztar:

--- a/package.yaml
+++ b/package.yaml
@@ -38,7 +38,6 @@ library:
     base: '>= 4.7 && < 5'
     bytestring: '>= 0.10.8 && < 0.11'
     directory: '>= 1.3 && < 1.4'
-    extra:
     filepath: '>= 1.4.1 && < 1.5'
     path: '>= 0.5 && < 0.8'
     # tar: '>= 0.5 && < 0.6'

--- a/src/Codec/Archive/ZTar/Zip.hs
+++ b/src/Codec/Archive/ZTar/Zip.hs
@@ -18,6 +18,7 @@ module Codec.Archive.ZTar.Zip
   ) where
 
 import qualified Codec.Archive.Zip as Zip
+import Control.Monad.Extra (concatMapM)
 import Control.Monad.IO.Class (liftIO)
 import Data.ByteString.Lazy (ByteString)
 import qualified Data.ByteString.Lazy as BS
@@ -57,21 +58,25 @@ create :: FilePath -- ^ archive to create
        -> IO ()
 create archive base paths = do
   archive' <- makeAbsolute archive
-  withCurrentDirectory base $ Zip.createArchive archive' $ mapM_ insert paths
+  withCurrentDirectory base $ do
+    files <- concatMapM search paths
+    Zip.createArchive archive' $ mapM_ insertFile files
   where
-    insert path = do
-      isFile <- liftIO $ doesFileExist path
-      isDir <- liftIO $ doesDirectoryExist path
+    search :: FilePath -> IO [FilePath]
+    search path = do
+      isFile <- doesFileExist path
+      isDir <- doesDirectoryExist path
       if
-        | isFile -> insertFile path
-        | isDir -> insertDir path
+        | isFile -> pure [path]
+        | isDir -> searchDir path
         | otherwise -> fail $ "Path does not exist: " ++ path
+    searchDir :: FilePath -> IO [FilePath]
+    searchDir path =
+      let mkPath = if path == "." then id else (path </>)
+      in concatMapM (search . mkPath) =<< liftIO (listDirectory path)
     insertFile path = do
       path' <- Zip.mkEntrySelector path
       Zip.loadEntry Zip.BZip2 path' path
-    insertDir path =
-      let mkPath = if path == "." then id else (path </>)
-      in mapM_ (insert . mkPath) =<< liftIO (listDirectory path)
 
 -- | Extract all the files contained in an archive compressed with Zip.
 --

--- a/src/Codec/Archive/ZTar/Zip.hs
+++ b/src/Codec/Archive/ZTar/Zip.hs
@@ -18,7 +18,7 @@ module Codec.Archive.ZTar.Zip
   ) where
 
 import qualified Codec.Archive.Zip as Zip
-import Control.Monad.Extra (concatMapM)
+import Control.Monad (liftM)
 import Control.Monad.IO.Class (liftIO)
 import Data.ByteString.Lazy (ByteString)
 import qualified Data.ByteString.Lazy as BS
@@ -77,6 +77,7 @@ create archive base paths = do
     insertFile path = do
       path' <- Zip.mkEntrySelector path
       Zip.loadEntry Zip.BZip2 path' path
+    concatMapM f = liftM concat . mapM f
 
 -- | Extract all the files contained in an archive compressed with Zip.
 --

--- a/src/Codec/Archive/ZTar/Zip.hs
+++ b/src/Codec/Archive/ZTar/Zip.hs
@@ -73,7 +73,7 @@ create archive base paths = do
     searchDir :: FilePath -> IO [FilePath]
     searchDir path =
       let mkPath = if path == "." then id else (path </>)
-      in concatMapM (search . mkPath) =<< liftIO (listDirectory path)
+      in concatMapM (search . mkPath) =<< listDirectory path
     insertFile path = do
       path' <- Zip.mkEntrySelector path
       Zip.loadEntry Zip.BZip2 path' path

--- a/test/Test.hs
+++ b/test/Test.hs
@@ -3,7 +3,6 @@
 import Codec.Archive.ZTar
 import Control.Monad (forM, forM_)
 import qualified Data.ByteString as BS
-import Data.ByteString.Arbitrary (ArbByteString(..))
 import Data.List (isPrefixOf, nub)
 import Data.Maybe (fromJust)
 import Path
@@ -22,6 +21,7 @@ import Path
 import Path.IO (doesFileExist, ensureDir, isLocationOccupied, withTempDir)
 import qualified System.FilePath.Windows as Windows
 import Test.QuickCheck
+import Test.QuickCheck.Instances.ByteString ()
 import Test.QuickCheck.Monadic
 import Test.Tasty (defaultMain, testGroup)
 import Test.Tasty.QuickCheck (testProperty)
@@ -103,7 +103,7 @@ arbitraryFileTree = mkFileTree Nothing (0 :: Int)
         isFile <- frequency [(1, pure False), (2^depth, pure True)]
         if isFile
           then do
-            Blind (ABS contents) <- arbitrary
+            Blind contents <- arbitrary
             return [(dir `maybeSlash` toRelFile path, contents)]
           else mkFileTree (Just $ dir `maybeSlash` toRelDir path) (depth + 1)
 

